extensions/socket/internal.m

```diff
@@ -0,0 +1,657 @@
+#import <LuaSkin/LuaSkin.h>
+#import <CocoaAsyncSocket/GCDAsyncSocket.h>
+#import "socket.h"
+
+// Definitions
+@interface HSAsyncSocket : GCDAsyncSocket
+@property int readCallback;
+@property int writeCallback;
+@property int connectCallback;
+@property NSTimeInterval timeout;
+@property NSMutableArray* connectedSockets;
+@end
+
+// Userdata for hs.socket objects
+static const char *USERDATA_TAG = "hs.socket";
+#define getUserData(L, idx) (__bridge HSAsyncSocket *)((asyncSocketUserData *)lua_touserdata(L, idx))->asyncSocket;
+
+
+// Delegate implementation
+@implementation HSAsyncSocket
+
+- (id)init {
+    self.readCallback = LUA_NOREF;
+    self.writeCallback = LUA_NOREF;
+    self.connectCallback = LUA_NOREF;
+    self.timeout = -1;
+    self.connectedSockets = [[NSMutableArray alloc] init];
+    return [super initWithDelegate:self delegateQueue:dispatch_get_main_queue()];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didConnectToHost:(NSString *)host port:(UInt16)port {
```

Comment:
Yes. You can init with `[super initWithDelegate:self delegateQueue:NULL]` and GCDAsyncSocket will create its own serial queues, or specify your own serial queue like:
```objc
- (id)init {
    dispatch_queue_t socketQueue = dispatch_queue_create("socketQueue", NULL);
...
    return [super initWithDelegate:self delegateQueue:socketQueue];
}
```
Then, in each of the delegate methods that operate on the UI/Lua in any way, you must wrap those operations like so:
```objc
    dispatch_async(dispatch_get_main_queue(), ^{ 
        @autoreleasepool {
            LuaSkin *skin = [LuaSkin shared];
            [skin logInfo:@"Data written to socket"];
            // do some lua callback or something
        }
    });
```
Failing to `dispatch_async` to the main thread will indeed crash as you say. I initially chose `dispatch_get_main_queue` as the `delegateQueue` because it 'just worked', but it does seem a bit more performant if I use separate queues and call `dispatch_async` to the main thread as shown above, presumably because GCDAsyncSocket avoids doing all of its internal operations on the main thread, waiting until the results are in. I have an uncommitted implementation of this and wanted your opinion before really putting it in.